ibis/expr/operations/core.py

```diff
@@ -75,11 +75,17 @@ def __coerce__(
         if isinstance(value, Value):
             return value
 
-        try:
+        if T is dt.Integer:
+            dtype = dt.infer(int(value))
+        elif T is dt.Floating:
+            dtype = dt.infer(float(value))
+        else:
```

Comment:
There are actually two branches handled by the `else`, when `T` is `None` in which case we could just use `dt.infer()` and when `T` is a `UnionType` e.g. `Value[dt.Numeric | dt.Boolean]`. The latter is used at just a couple of cases, so we could actually remove support for that.